src/storage/v2/constraints/type_constraints.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright 2024 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "type_constraints.hpp"
+#include <optional>
+#include "storage/v2/constraints/type_constraints_type.hpp"
+#include "storage/v2/id_types.hpp"
+#include "storage/v2/property_value.hpp"
+#include "utils/algorithm.hpp"
+
+namespace memgraph::storage {
+
+[[nodiscard]] std::optional<ConstraintViolation> TypeConstraints::ValidateVertexOnConstraint(const Vertex &vertex,
+                                                                                             LabelId label,
+                                                                                             PropertyId property,
+                                                                                             TypeConstraintsType type) {
+  if (vertex.deleted || !utils::Contains(vertex.labels, label)) return std::nullopt;
+
+  auto prop_value = vertex.properties.GetProperty(property);
+  if (prop_value.IsNull()) {
+    return std::nullopt;
+  }
+
+  if (PropertyValueToTypeConstraintType(prop_value) != type) {
+    return ConstraintViolation{ConstraintViolation::Type::TYPE, label, type, std::set<PropertyId>{property}};
+  }
+
+  return std::nullopt;
+}
+
+[[nodiscard]] std::optional<ConstraintViolation> TypeConstraints::Validate(const Vertex &vertex) const {
+  if (vertex.deleted) return std::nullopt;
+
+  for (auto const &[property_id, property_value] : vertex.properties.Properties()) {
```

Comment:
Nope even worse IMO, this forces hydration of property values from property_store -> property_value even when they are not needed.

- We want to build a set of property type constraints we expect.
- Only do it for the labels that the vertex has
- Do the type checking inside the property store without hydrating a property value at all

Ideally we could do this before inserting into the property store. ie. Check it is ok to set rather than validating that the set was ok.